Sources/S3/Extensions/S3+Private.swift

```diff
@@ -1,28 +1,55 @@
-//
-//  S3+Private.swift
-//  S3
-//
-//  Created by Ondrej Rafaj on 19/04/2018.
-//
-
 import Foundation
 import Vapor
 import HTTP
 
-
 extension S3 {
     
     /// Make an S3 request
-    func make(request url: URL, method: HTTPMethod, headers: HTTPHeaders, data: Data? = nil, on container: Container) throws -> Future<Response> {
-        let client = try container.make(Client.self)
-        let request = Request(using: container)
-        request.http.method = method
-        request.http.headers = headers
+    func make(request url: URL, method: HTTPMethod, headers: HTTPHeaders, data: Data? = nil, cachePolicy: URLRequest.CachePolicy = .useProtocolCachePolicy, on container: Container) throws -> Future<Response> {
+        var request = URLRequest(url: url, cachePolicy: cachePolicy)
+        request.httpMethod = method.string
+        request.httpBody = data ?? Data()
+        headers.forEach { key, val in
+            request.addValue(val, forHTTPHeaderField: key.description)
+        }
+        
+        return execute(request, on: container)
+    }
+    
+    /// Execute given request with URLSession.shared
+    func execute(_ request: URLRequest, on container: Container) -> Future<Response> {
+        let promise = container.eventLoop.newPromise(Response.self)
+        
+        URLSession.shared.dataTask(with: request, completionHandler: { (data, urlResponse, error) in
+            if let error = error {
```

Comment:
is there nay situation when we wouldn't hit the complete closure? (just asking) ... if there is'd have a future leak and NIO would crash

Reply:
Why would it crash? What leak?
Well it is Foundation and it is the only closure I can pass in the call so I suppose it is called every time :)

Reply:
I am just asking ... NIO is super strict about having their futures fulfilled and the whole thing will crash on purpose if they aint

Reply:
Well, the `FoundationClient` is doing the same so I hope it is ok.